# Fase 1: Análisis Exploratorio de Datos(EDA).

Antes de llevar a cabo el proyecto es crucial comprender mejor el conjunto de datos y sus características. Para ello deberás hacer un análisis exploratorio detallado del conjunto de datos para familiarizarte con ellos y entender que información tenemos.

# 📊 Preguntas clave para el análisis de rotación

| Nº | Pregunta | Variables involucradas | Objetivo del análisis |
|----|----------|------------------------|------------------------|
| 1 | ¿Qué porcentaje de empleados abandona la empresa? | Attrition | Medir la rotación general. |
| 2 | ¿La rotación cambia según departamento? | Department, Attrition | Comparar rotación por áreas. |
| 3 | ¿Qué puestos tienen más riesgo de rotación? | JobRole, Attrition | Identificar roles con mayor salida. |
| 4 | ¿La satisfacción laboral influye en la salida? | JobSatisfaction, EnvironmentSatisfaction, Attrition | Evaluar impacto de la satisfacción. |
| 5 | ¿Las horas extra están relacionadas con mayor rotación? | OverTime, Attrition | Ver si trabajar más horas aumenta la salida. |
| 6 | ¿La distancia desde casa afecta a la retención? | DistanceFromHome, Attrition | Analizar si vivir lejos influye en irse. |
| 7 | ¿El salario mensual influye en la rotación? | MonthlyIncome, Attrition | Comparar ingresos entre quienes se quedan y se van. |
| 8 | ¿La antigüedad reduce o aumenta la rotación? | YearsAtCompany, TotalWorkingYears, YearsInCurrentRole | Ver cómo influye la experiencia y permanencia. |
| 9 | ¿La falta de promoción está relacionada con la salida? | YearsSinceLastPromotion, Attrition | Evaluar si no promocionar aumenta la rotación. |
| 10 | ¿Existe diferencia en la rotación según el género? | Gender, Attrition | Detectar si hay disparidad en la salida entre hombres y mujeres. |
| 11 | ¿Qué variables necesitan limpieza antes del análisis? | EmployeeCount, Over18, StandardHours + nulos/duplicados | Identificar columnas constantes y problemas de calidad. |


In [1]:
import pandas as pd
import numpy as np #Esto es para trabajar con arrays y operaciones matemáticas
import warnings #Esto sirve para manejar las advertencias que puedan surgir durante la ejecución del código
warnings.filterwarnings('ignore')

# Mostrar todas las columnas sin truncar
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 60)

print("✅ Librerías cargadas correctamente")


✅ Librerías cargadas correctamente


In [2]:
df = pd.read_csv('hr.csv')

print(f"El número de filas y columnas es: {df.shape[0]} filas y {df.shape[1]} columnas")

El número de filas y columnas es: 1474 filas y 35 columnas


In [3]:
df.head(5)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0
2,37.0,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,lABORATORY tECHNICIAN,3.0,Single,2090.0,2396,6,Y,Yes,15,3,2,NaN,0,7,3.0,3,0,0,0,0.0
3,33.0,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,rESEARCH sCIENTIST,3.0,Married,2909.0,23159,1,Y,Yes,11,3,3,80.0,0,8,3.0,3,8,7,3,0.0
4,27.0,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,lABORATORY tECHNICIAN,2.0,Married,3468.0,16632,9,Y,No,12,3,4,80.0,1,6,3.0,3,2,2,2,2.0


Cositas para cambiar:
- JobRole --> mal escrito
- MaritalStatus --> en alguno pone marreid en vez de married y nulos
- EducationField --> Error de escritura, espacios... arreglar con .strip()
- AGE está en formato float, tenemos que convertirla a int

Eliminar:
- EmployeeCount, Over18, StandardHours porque no aportan nada a nuestro análisis y ya veremos si tambien eliminamos EmployeeNumber (un número único para cada empleado, no tiene ningún valor predictivo).

Estudiar:
- Attrition (DESGASTE) es la columna que nos interesa analizar, es la variable objetivo, es decir, lo que queremos predecir. Es una variable categórica, tiene dos valores: Yes y No. Tenemos que convertirla a 0 y 1 para poder trabajar con ella.

**¿Qué observamos?**
- Hay columnas de datos personales (`Age`, `Gender`, `MaritalStatus`)
- Columnas laborales (`Department`, `JobRole`, `YearsAtCompany`)
- Columnas de satisfacción y rendimiento (`JobSatisfaction`, `PerformanceRating`, `WorkLifeBalance`)
- La variable objetivo: `Attrition` — si el empleado se fue (Yes) o se quedó (No)
- Columna `JobRole` con un formato de texto extraño (mayúsculas y minúsculas mezcladas)

In [4]:
df.info

<bound method DataFrame.info of        Age Attrition     BusinessTravel  DailyRate              Department  \
0     41.0       Yes      Travel_Rarely       1102                   Sales   
1     49.0        No  Travel_Frequently        279  Research & Development   
2     37.0       Yes      Travel_Rarely       1373  Research & Development   
3     33.0        No  Travel_Frequently       1392  Research & Development   
4     27.0        No      Travel_Rarely        591  Research & Development   
...    ...       ...                ...        ...                     ...   
1469  34.0        No      Travel_Rarely        628  Research & Development   
1470  28.0        No      Travel_Rarely        866                   Sales   
1471  53.0        No      Travel_Rarely       1084  Research & Development   
1472  24.0       Yes      Travel_Rarely        240         Human Resources   
1473  45.0        No      Travel_Rarely       1339  Research & Development   

      DistanceFromHome  Educati

**Problemas detectados:**

🔴 **Columnas con nulos** (Non-Null Count < 1.474):
| Columna | Nulos | Impacto para el proyecto |
|---|---|---|
| `YearsWithCurrManager` | 148 | Dato laboral importante |
| `MaritalStatus` | 132 | Dato personal |
| `BusinessTravel` | 117 | Dato laboral |
| `TrainingTimesLastYear` | 88 | Clave para el perfil candidato |
| `Age` | 73 | Dato básico |
| `EducationField` | 58 | Contexto del empleado |
| `OverTime` | 44 | Clave para el perfil candidato |
| `JobSatisfaction` | 29 | Clave para el perfil candidato |
| `Department` | 29 | Dato laboral |
| `MonthlyIncome` | 14 | Dato económico |

🔴 **Tipos incorrectos:**
- `Age`, `JobSatisfaction`, `MonthlyIncome`, `TrainingTimesLastYear`, `YearsWithCurrManager` aparecen como `float64` cuando deberían ser `int` — nadie tiene 32.0 años ni 3.0 niveles de satisfacción. Esto ocurre porque pandas convierte a float las columnas con nulos.

→ Todo esto se corregirá en la **Fase 2**.

In [5]:
## Análisis detallado de valores nulos

#Calculamos exactamente cuántos nulos hay en cada columna y qué porcentaje representan sobre el total. Un porcentaje alto de nulos puede indicar que esa columna no es fiable.

nulos = df.isnull().sum()
pct   = (nulos / len(df) * 100).round(2)

resumen_nulos = pd.DataFrame({
    'Nulos': nulos,
    'Porcentaje (%)': pct
}).sort_values('Nulos', ascending=False)

print("Columnas CON nulos:")
print(resumen_nulos[resumen_nulos['Nulos'] > 0].to_string())
print(f"\nColumnas SIN nulos: {(nulos == 0).sum()}")
print(f"Total de valores nulos en el dataset: {nulos.sum()}")

Columnas CON nulos:
                       Nulos  Porcentaje (%)
StandardHours            164           11.13
YearsWithCurrManager     148           10.04
MaritalStatus            132            8.96
BusinessTravel           117            7.94
TrainingTimesLastYear     88            5.97
Age                       73            4.95
EducationField            58            3.93
OverTime                  44            2.99
JobSatisfaction           29            1.97
Department                29            1.97
MonthlyIncome             14            0.95

Columnas SIN nulos: 24
Total de valores nulos en el dataset: 896


**Interpretación:**
- Ninguna columna supera el 12% de nulos, así que no hay que descartar ninguna.
- `StandardHours` tiene nulos pero como es una columna redundante se eliminará igualmente.
- Los nulos se imputarán en la Fase 2: **mediana** para columnas numéricas (más robusta que la media ante valores extremos) y **moda** para columnas categóricas (el valor más frecuente).


## Pregunta 1 - ¿Qué porcentaje de empleados abandona la empresa? 

`Attrition` es la columna más importante del dataset. Indica si el empleado abandonó la empresa (`Yes`) o sigue en ella (`No`). 

Para el proyecto de cursos de idiomas, **solo nos interesan los empleados con `Attrition = No`**. Ofrecer un curso caro a alguien que ya se fue o que tiene alta probabilidad de irse sería un desperdicio de recursos.

In [6]:
print("=== DISTRIBUCIÓN DE ATTRITION ===\n")
print(df['Attrition'].value_counts())
print()
print("Porcentaje:")
print((df['Attrition'].value_counts(normalize=True) * 100).round(1).astype(str) + '%')
print()

# Universo de candidatos potenciales
candidatos = df[df['Attrition'] == 'No']
print(f"━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
print(f"Total empleados en el dataset:       {len(df):>5}")
print(f"Empleados que se fueron (excluidos): {df['Attrition'].eq('Yes').sum():>5}")
print(f"Candidatos potenciales al curso:     {len(candidatos):>5}")
print(f"━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")


=== DISTRIBUCIÓN DE ATTRITION ===

Attrition
No     1236
Yes     238
Name: count, dtype: int64

Porcentaje:
Attrition
No     83.9%
Yes    16.1%
Name: proportion, dtype: str

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Total empleados en el dataset:        1474
Empleados que se fueron (excluidos):   238
Candidatos potenciales al curso:      1236
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━


**Interpretación:**
- El **16,1%** de los empleados se fue de la empresa — esto es la tasa de rotación.
- El **83,9%** sigue en la empresa → **1.236 candidatos potenciales** antes de aplicar cualquier otro filtro.
- A partir de aquí, todos los análisis de perfil se harán sobre este subconjunto.

## PREGUNTA 2. LA ROTACION CAMBIA SEGÚN DEPARTAMENTO?

Contar departamentos de quienes se fueron (filtro empleados que se fueron y miro qué departamentos aparecen más)

In [7]:
df[df["Attrition"] == "Yes"]["Department"].value_counts().reset_index()

,Department,count
0,Research & Development,126
1,Sales,89
2,Human Resources,13


**Interpretación**
- El departamento de ventas es el que más rotación tiene, seguido de investigación y desarrollo y luego recursos humanos.
Esto puede indicar que los empleados de ventas tienen más probabilidades de irse, lo que podría ser un área de preocupación para la empresa.

### PREGUNTA3. ¿Qué puestos tienen más riesgo de rotación?  JobRole, Attrition  Identificar roles con mayor salida. 
 
- Ver todos los puestos con más rotación, para saber cuales tiene más tasa de abandono:

In [8]:
#para calcular la tasa de rotación por puesto, necesitamos el número de empleados que se fueron por cada puesto (absolutos) y el número total de empleados por cada puesto (total).
absolutos = df[df["Attrition"] == "Yes"]["JobRole"].value_counts().reset_index()
absolutos.columns = ["JobRole", "Bajas"]

# Total de empleados por puesto
total = df["JobRole"].value_counts().reset_index()
total.columns = ["JobRole", "Total"]

# Unir y calcular tasa de rotación
rotacion = absolutos.merge(total, on="JobRole")
rotacion["Tasa_%"] = (rotacion["Bajas"] / rotacion["Total"] * 100).round(1)
rotacion = rotacion.sort_values("Tasa_%", ascending=False)

rotacion

,JobRole,Bajas,Total,Tasa_%
3,sALES rEPRESENTATIVE,33,83,39.8
4,hUMAN rESOURCES,13,53,24.5
0,lABORATORY tECHNICIAN,62,259,23.9
1,sALES eXECUTIVE,57,327,17.4
2,rESEARCH sCIENTIST,47,293,16.0
6,hEALTHCARE rEPRESENTATIVE,9,131,6.9
5,mANUFACTURING dIRECTOR,10,146,6.8
7,mANAGER,5,102,4.9
8,rESEARCH dIRECTOR,2,80,2.5


**Interpretaciones clave**

- Sales Representative es el puesto más crítico — casi el 40% abandona la empresa. Es un rol orientado al cliente y clave para la expansión europea, lo que lo hace doblemente preocupante.
- Human Resources y Laboratory Technician también superan el 20% de rotación — niveles muy altos que merecen atención.
- Research Director y Manager son los puestos más estables, con tasas por debajo del 5% — perfiles senior con mayor vinculación a la empresa.
- Para el análisis del curso de idiomas: tiene poco sentido invertir en Sales Representative si casi el 40% se va. Habría que cruzarlo con otros filtros.

### PREGUNTA 4: ¿La satisfacción laboral influye en la salida?  JobSatisfaction, EnvironmentSatisfaction, Attrition  Evaluar impacto de la satisfacción. 

DESORDENADO, DAR UNA VUELTA!!

In [10]:
# Ver satisfacción de quienes se fueron:
df[df["Attrition"] == "Yes"]["JobSatisfaction"].value_counts().reset_index()

,JobSatisfaction,count
0,3.0,71
1,1.0,64
2,4.0,51
3,2.0,46


In [15]:
#dato en porcentaje:
df[df["Attrition"] == "Yes"]["JobSatisfaction"].value_counts(normalize=True) * 100

JobSatisfaction
3.0    30.603448
1.0    27.586207
4.0    21.982759
2.0    19.827586
Name: proportion, dtype: float64

In [16]:
df[df["JobSatisfaction"].isnull()]["Attrition"].value_counts() 
#esta información es importante porque nos dice que los nulos de JobSatisfaction corresponden a empleados que se fueron, 
# por lo que no podemos imputar un valor de satisfacción a esos casos. Por lo tanto, lo más adecuado es eliminar esas filas para evitar introducir sesgos en el análisis.

Attrition
No     23
Yes     6
Name: count, dtype: int64

In [17]:
# PREGUNTA 4. ¿LA SATISFACCIÓN LABORAL INFLUYE EN LA SALIDA? (combinado con lo anterior, para ver si hay relación entre puesto, satisfacción y rotación)

result = df.groupby("Attrition")["JobSatisfaction"].value_counts(normalize=True).mul(100).round(1).unstack()
result.columns = [f"Nivel {int(c)}" for c in result.columns]

result

,Nivel 1,Nivel 2,Nivel 3,Nivel 4
Attrition,,,,
No,18.3,18.9,29.5,33.3
Yes,27.6,19.8,30.6,22.0


In [18]:
# PREGUNTA 4. ¿LA SATISFACCIÓN LABORAL INFLUYE EN LA SALIDA?

df.groupby("Attrition")[["JobSatisfaction", "EnvironmentSatisfaction"]].mean().round(2)

,JobSatisfaction,EnvironmentSatisfaction
Attrition,,
No,2.78,2.77
Yes,2.47,2.47


**Conclusiones:**
1. Los que se van tienen menor satisfacción en ambas métricas
Tanto en satisfacción laboral como en el entorno, los empleados que abandonan puntúan consistentemente por debajo de los que se quedan.
2. La diferencia no es dramática pero es consistente
~0.3 puntos de diferencia en ambas variables. No es un abismo, pero la consistencia entre las dos métricas refuerza que la satisfacción sí es un factor que influye.
3. Ningún grupo está realmente satisfecho
La escala es del 1 al 4 y los que se quedan solo puntúan 2.78 — esto indica un problema general de satisfacción en la empresa, no solo entre los que se van.

### PREGUNTA 5:¿Las horas extra están relacionadas con mayor rotación? | OverTime, Attrition | Ver si trabajar más horas aumenta la salida. 

In [19]:
# PREGUNTA 5. ¿LAS HORAS EXTRA ESTÁN RELACIONADAS CON MAYOR ROTACIÓN?

overtime = df.groupby("OverTime")["Attrition"].value_counts().unstack(fill_value=0)
overtime["Tasa de Rotación (%)"] = (overtime["Yes"] / (overtime["Yes"] + overtime["No"]) * 100).round(1)
overtime = overtime.reset_index()
overtime.columns = ["Horas Extra", "Se quedaron", "Se fueron", "Tasa de Rotación (%)"]

overtime

,Horas Extra,Se quedaron,Se fueron,Tasa de Rotación (%)
0,No,920,105,10.2
1,Yes,280,125,30.9


**Interpretaciones:**
1. Las horas extra triplican la rotación
Empleados con horas extra tienen una tasa del 30.9% frente al 10.2% de los que no las hacen.
2. Es una de las variables más impactantes del dataset
Una diferencia de 20 puntos porcentuales es muy significativa — pocas variables tendrán tanto peso como esta.
3. Coste oculto para ABC Corporation
Las horas extra pueden parecer productivas a corto plazo, pero si 1 de cada 3 empleados acaba marchándose, el coste real de reemplazarlos supera con creces el beneficio.
4. Conecta con la pregunta anterior
Más horas → menos equilibrio trabajo-vida → menor satisfacción → mayor rotación. Es una cadena de causa-efecto que refuerza los resultados de la pregunta 4.

### PREGUNTA 6: ¿La distancia desde casa afecta a la retención? | DistanceFromHome, Attrition | Analizar si vivir lejos influye en irse. 

In [20]:
# PREGUNTA 6. ¿LA DISTANCIA DESDE CASA AFECTA A LA RETENCIÓN?

df.groupby("Attrition")["DistanceFromHome"].mean().round(2)

Attrition
No      8.91
Yes    10.68
Name: DistanceFromHome, dtype: float64

**Conclusiones:**
1. Los que se van viven más lejos
Una diferencia de casi 2 km de media — los que abandonan viven a 10.68 km frente a los 8.91 de los que se quedan.
2. El impacto existe pero es moderado
No es una diferencia dramática. La distancia influye pero probablemente como factor acumulativo junto a otros como horas extra o baja satisfacción, no por sí sola.
3. La mediana es más reveladora que la media
La mediana de los que se van es 9 km frente a 7 km de los que se quedan — confirma el patrón sin que los valores extremos distorsionen.

### PREGUNTA 7. ¿El salario mensual influye en la rotación? | MonthlyIncome, Attrition | Comparar ingresos entre quienes se quedan y se van. |

In [21]:
# PREGUNTA 7. ¿EL SALARIO MENSUAL INFLUYE EN LA ROTACIÓN?

df.groupby("Attrition")["MonthlyIncome"].describe().round(2)

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
No,1226.0,6832.69,4813.70,1051.0,3211.25,5205.0,8836.25,19999.0
Yes,234.0,4741.03,3628.36,1009.0,2366.25,3187.0,5824.25,19859.0


**Conclusiones:**
1. El salario es una de las variables más impactantes
Los que se van cobran de media 2.091$ menos al mes que los que se quedan. Una diferencia del 30% — muy significativa.
2. La mediana lo confirma sin distorsión
La mediana de los que se van es 3.187$ frente a 5.205$ de los que se quedan. Casi 2.000$ de diferencia — el patrón es claro y consistente.
3. El salario bajo es un factor de riesgo claro
Junto con las horas extra y la baja satisfacción, el salario bajo forma el núcleo de los factores que empujan a los empleados a irse.
4. Implicación para ABC Corporation
Revisar la política salarial, especialmente en los puestos con mayor rotación como Sales Representative y Laboratory Technician, podría tener un impacto directo en la retención.

### PREGUNTA 8: ¿La antigüedad reduce o aumenta la rotación? | YearsAtCompany, TotalWorkingYears, YearsInCurrentRole | Ver cómo influye la experiencia y permanencia. |

In [22]:
df.groupby("Attrition")[["YearsAtCompany", "TotalWorkingYears", "YearsInCurrentRole"]].mean().round(2)

,YearsAtCompany,TotalWorkingYears,YearsInCurrentRole
Attrition,,,
No,7.36,11.86,4.48
Yes,5.11,8.21,2.89


**Conclusiones:**

1. Los que se van llevan menos tiempo en la empresa
5.11 años de media frente a 7.36 — los empleados más nuevos son los más vulnerables a irse.
2. El patrón se repite en las tres variables
Años en empresa, años totales trabajados y años en el rol actual — en las tres los que se van puntúan consistentemente más bajo. Es una señal muy robusta.
3. Los primeros años son los más críticos
Con menos de 5 años en la empresa el riesgo de abandono es mayor. ABC Corporation debería poner especial atención en la retención de empleados recientes.
4. Años en el rol actual es llamativo
Solo 2.89 años de media entre los que se van — puede indicar que la falta de rotación interna o promoción dentro del rol también contribuye a la salida, lo que conecta directamente con la pregunta 9.

### PREGUNTA 9:  ¿La falta de promoción está relacionada con la salida? | YearsSinceLastPromotion, Attrition | Evaluar si no promocionar aumenta la rotación. |

In [23]:
df.groupby("Attrition")["YearsSinceLastPromotion"].describe().round(2)

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
No,1236.0,2.23,3.23,0.0,0.0,1.0,3.0,15.0
Yes,238.0,1.94,3.15,0.0,0.0,1.0,2.0,15.0


**Conclusiones:**
1. Sorpresa: los que se van llevan menos tiempo sin promoción
La diferencia es mínima (2.23 vs 1.94) y va en sentido contrario a lo esperado — los que se van fueron promocionados más recientemente.
2. La falta de promoción no explica la rotación por sí sola
A diferencia de las preguntas anteriores, aquí no hay un patrón claro. La promoción reciente no garantiza que el empleado se quede.
3. Puede haber otros factores detrás
Una promoción sin aumento salarial o sin mejora real de condiciones podría no ser suficiente para retener. Conecta con la pregunta 7 — el salario pesa más que el título.
4. Variable menos prioritaria para el análisis
De todas las variables analizadas hasta ahora, esta es la que menos poder explicativo tiene sobre la rotación.

### PREGUNTA 10: ¿Existe diferencia en la rotación según el género? | Gender, Attrition | Detectar si hay disparidad en la salida entre hombres y mujeres.

In [29]:
# PREGUNTA 10. ¿EXISTE DIFERENCIA EN LA ROTACIÓN SEGÚN EL GÉNERO?

genero = df.groupby("Gender")["Attrition"].value_counts().unstack(fill_value=0)
genero["Tasa_%"] = (genero["Yes"] / (genero["Yes"] + genero["No"]) * 100).round(1)
genero.columns = ["Se quedaron", "Se fueron", "Tasa_%"]

genero

,Se quedaron,Se fueron,Tasa_%
Gender,,,
Female,502,87,14.8
Male,734,151,17.1


**Conclusiones:**
1. Los hombres tienen una tasa de rotación ligeramente mayor
17.1% frente al 14.8% de las mujeres — una diferencia de 2.3 puntos porcentuales.
2. La diferencia existe pero no es alarmante
No hay una disparidad grave por género. Ambos grupos tienen tasas relativamente similares, lo que sugiere que el género por sí solo no es un factor determinante.
3. Convendría cruzarlo con otras variables
Para entender mejor si hay algún patrón oculto, sería interesante cruzar género WorkLifeBalance y Attrition

In [30]:
# PREGUNTA 10 CRUZADA CON WORKLIFEBALANCE

resultado = df.groupby(["Gender", "WorkLifeBalance"])["Attrition"].value_counts(normalize=True).mul(100).round(1).unstack(fill_value=0)
resultado.columns = ["Se quedaron (%)", "Se fueron (%)"]

resultado

Se quedaron (%)  Se fueron (%)
Gender WorkLifeBalance                                
Female 1                           76.7           23.3
       2                           86.0           14.0
       3                           87.2           12.8
       4                           75.4           24.6
Male   1                           64.0           36.0
       2                           81.2           18.8
       3                           84.7           15.3
       4                           86.5           13.5

Conclusiones:
1. Los hombres con mal balance son los más vulnerables
Un 36% de hombres con WorkLifeBalance nivel 1 abandona — el dato más alto de toda la tabla.
2. Las mujeres con balance excelente también se van más
Sorprendentemente, las mujeres con nivel 4 tienen un 24.6% de rotación, mayor que los niveles 2 y 3. Puede merecer una investigación más profunda.
3. El nivel 3 es el punto óptimo de retención para ambos géneros
Tanto hombres (15.3%) como mujeres (12.8%) tienen su menor tasa de rotación en el nivel 3, no en el 4.
4. El género sí marca diferencia cuando se combina con el balance
El cruce revela patrones que no se veían analizando cada variable por separado — exactamente el valor de combinar variables.

**Conclusiones 2.0**
1. Los hombres con mal balance son el grupo más crítico
Con un 36% de rotación en nivel 1, los hombres con peor equilibrio vida-trabajo son el segmento más vulnerable de toda la empresa. ABC Corporation debería priorizar políticas de flexibilidad especialmente dirigidas a este perfil.
2. El nivel 3 retiene mejor que el nivel 4
Contraintuitivamente, el balance "muy bueno" (nivel 3) retiene más que el "excelente" (nivel 4) en ambos géneros. Esto sugiere que los empleados con balance perfecto pueden ser perfiles con menos carga de trabajo, menor implicación o menor ambición profesional — y por tanto más propensos a buscar nuevos retos.
3. Las mujeres con balance excelente son una anomalía a investigar
Un 24.6% de rotación en mujeres con nivel 4 rompe el patrón esperado. Podría indicar que para las mujeres el equilibrio vida-trabajo no es suficiente si no va acompañado de otros factores como salario, promoción o reconocimiento.
4. El género solo cuenta parte de la historia
Analizar género de forma aislada puede llevar a conclusiones superficiales. Es cuando se cruza con variables como el balance, el salario o las horas extra cuando aparecen los patrones realmente accionables para la toma de decisiones.

### PREGUNTA 11: ¿Qué variables necesitan limpieza antes del análisis? | EmployeeCount, Over18, StandardHours + nulos/duplicados | Identificar columnas constantes y problemas de calidad. |

In [38]:
# Columnas constantes
constantes = [col for col in df.columns if df[col].nunique() == 1]
pd.DataFrame({
    "Columna": constantes,
    "Valor único": [df[col].unique()[0] for col in constantes],
    "Acción": ["Eliminar en Fase 2"] * len(constantes)
})

,Columna,Valor único,Acción
0,EmployeeCount,1,Eliminar en Fase 2
1,Over18,Y,Eliminar en Fase 2
2,StandardHours,80.0,Eliminar en Fase 2


In [35]:
# Nulos por columna ordenados
df.isnull().sum()[df.isnull().sum() > 0].sort_values(ascending=False)


StandardHours            164
YearsWithCurrManager     148
MaritalStatus            132
BusinessTravel           117
TrainingTimesLastYear     88
Age                       73
EducationField            58
OverTime                  44
Department                29
JobSatisfaction           29
MonthlyIncome             14
dtype: int64

In [36]:
# Duplicados
df.duplicated().sum()

np.int64(4)

In [39]:
def eda_basico(df):
    print("🔍 Primeras filas del DataFrame:")
    display(df.head())

    print("📐 Dimensiones:")
    print(f"{df.shape[0]} filas y {df.shape[1]} columnas\n")

    print("🧠 Información general:")
    display(df.info())

    print("📊 Tipos de datos por columna:")
    display(df.dtypes.rename("Tipo").to_frame())

    print("📉 Descripción de columnas numéricas:")
    display(df.describe())

    print("🔤 Descripción de columnas categóricas:")
    display(df.describe(include=["O"]))

    print("🚫 Valores nulos por columna:")
    nulos = df.isnull().sum()
    nulos = nulos[nulos > 0].sort_values(ascending=False)
    if len(nulos) > 0:
        display(pd.DataFrame({
            "Nulos": nulos,
            "Porcentaje (%)": (nulos / len(df) * 100).round(2)
        }))
    else:
        print("No hay valores nulos.\n")

    print("🗑️ Columnas constantes:")
    constantes = [col for col in df.columns if df[col].nunique() == 1]
    if constantes:
        display(pd.DataFrame({
            "Columna": constantes,
            "Valor único": [df[col].unique()[0] for col in constantes],
            "Acción": ["Eliminar en Fase 2"] * len(constantes)
        }))
    else:
        print("No hay columnas constantes.\n")

    print("📎 Filas duplicadas:")
    dup_count = df.duplicated().sum()
    print(f"Duplicadas: {dup_count}")
    if dup_count > 0:
        print("Ejemplo de duplicados:")
        display(df[df.duplicated()].head())
    else:
        print("No hay filas duplicadas.\n")

In [40]:
eda_basico(df)

🔍 Primeras filas del DataFrame:


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0
2,37.0,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,lABORATORY tECHNICIAN,3.0,Single,2090.0,2396,6,Y,Yes,15,3,2,NaN,0,7,3.0,3,0,0,0,0.0
3,33.0,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,rESEARCH sCIENTIST,3.0,Married,2909.0,23159,1,Y,Yes,11,3,3,80.0,0,8,3.0,3,8,7,3,0.0
4,27.0,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,lABORATORY tECHNICIAN,2.0,Married,3468.0,16632,9,Y,No,12,3,4,80.0,1,6,3.0,3,2,2,2,2.0


📐 Dimensiones:
1474 filas y 35 columnas

🧠 Información general:
<class 'pandas.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1401 non-null   float64
 1   Attrition                 1474 non-null   str    
 2   BusinessTravel            1357 non-null   str    
 3   DailyRate                 1474 non-null   int64  
 4   Department                1445 non-null   str    
 5   DistanceFromHome          1474 non-null   int64  
 6   Education                 1474 non-null   int64  
 7   EducationField            1416 non-null   str    
 8   EmployeeCount             1474 non-null   int64  
 9   EmployeeNumber            1474 non-null   int64  
 10  EnvironmentSatisfaction   1474 non-null   int64  
 11  Gender                    1474 non-null   str    
 12  HourlyRate                1474 non-null   int64  
 13  JobInvolve

None

📊 Tipos de datos por columna:


,Tipo
Age,float64
Attrition,str
BusinessTravel,str
DailyRate,int64
Department,str
DistanceFromHome,int64
Education,int64
EducationField,str
EmployeeCount,int64
EmployeeNumber,int64


📉 Descripción de columnas numéricas:


,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,JobSatisfaction,MonthlyIncome,MonthlyRate,NumCompaniesWorked,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1401.000000,1474.000000,1474.000000,1474.000000,1474.0,1474.000000,1474.000000,1474.000000,1474.000000,1474.000000,1445.000000,1460.000000,1474.000000,1474.000000,1474.000000,1474.000000,1474.000000,1310.0,1474.000000,1474.000000,1386.000000,1474.000000,1474.000000,1474.000000,1474.000000,1326.000000
mean,36.944325,802.702171,9.199457,2.911126,1.0,1024.471506,2.723881,65.887381,2.729986,2.063772,2.728720,6497.448630,14328.611940,2.688602,15.206920,3.153324,2.713704,80.0,0.794437,11.274084,2.797980,2.761872,6.996608,4.221845,2.183175,4.061840
std,9.105009,403.539530,8.104266,1.024267,0.0,602.421193,1.093280,20.310444,0.712861,1.106055,1.106409,4706.059294,7120.225285,2.496443,3.659274,0.360422,1.080338,0.0,0.851784,7.785972,1.296882,0.705625,6.122920,3.622032,3.219356,3.517603
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,1.000000,1009.000000,2094.000000,0.000000,11.000000,3.000000,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,488.750000,2.000000,48.000000,2.000000,1.000000,2.000000,2909.000000,8054.500000,1.000000,12.000000,3.000000,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,803.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,3.000000,4907.000000,14248.500000,2.000000,14.000000,3.000000,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,4.000000,8377.000000,20465.750000,4.000000,18.000000,3.000000,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,4.000000,19999.000000,26999.000000,9.000000,25.000000,4.000000,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


🔤 Descripción de columnas categóricas:


,Attrition,BusinessTravel,Department,EducationField,Gender,JobRole,MaritalStatus,Over18,OverTime
count,1474,1357,1445,1416,1474,1474,1342,1474,1430
unique,2,3,3,6,2,9,4,1,2
top,No,Travel_Rarely,Research & Development,Life Sciences,Male,sALES eXECUTIVE,Married,Y,No
freq,1236,955,941,582,885,327,604,1474,1025


🚫 Valores nulos por columna:


,Nulos,Porcentaje (%)
StandardHours,164,11.13
YearsWithCurrManager,148,10.04
MaritalStatus,132,8.96
BusinessTravel,117,7.94
TrainingTimesLastYear,88,5.97
Age,73,4.95
EducationField,58,3.93
OverTime,44,2.99
Department,29,1.97
JobSatisfaction,29,1.97


🗑️ Columnas constantes:


,Columna,Valor único,Acción
0,EmployeeCount,1,Eliminar en Fase 2
1,Over18,Y,Eliminar en Fase 2
2,StandardHours,80.0,Eliminar en Fase 2


📎 Filas duplicadas:
Duplicadas: 4
Ejemplo de duplicados:


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1470,28.0,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,4,Male,84,3,2,sALES eXECUTIVE,1.0,Single,8463.0,23490,0,Y,No,18,3,4,NaN,0,6,4.0,3,5,4,1,NaN
1471,53.0,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,4,Female,57,4,2,mANUFACTURING dIRECTOR,1.0,Divorced,4450.0,26250,1,Y,No,11,3,3,NaN,2,5,3.0,3,4,2,1,3.0
1472,24.0,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,hUMAN rESOURCES,3.0,Married,1555.0,11585,1,Y,No,11,3,3,80.0,1,1,2.0,3,1,0,0,0.0
1473,45.0,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,rESEARCH sCIENTIST,1.0,Divorced,9724.0,18787,2,Y,No,17,3,3,NaN,1,25,2.0,3,1,0,0,0.0
